modules/currency-l1/src/main/scala/io/constellationnetwork/currency/l1/http/DataApplicationRoutes.scala

```diff
@@ -47,55 +47,103 @@ final case class DataApplicationRoutes[F[_]: Async: Hasher: SecurityProvider: L1
+            }
+          case None =>
+            InternalServerError(CurrencySnapshotUnavailable.toApplicationError)
+        }
+
+      case None =>
+        InternalServerError(GL0SnapshotOrdinalUnavailable.toApplicationError)
     }
 
   protected val public: HttpRoutes[F] = HttpRoutes.of[F] {
     case req @ POST -> Root / "data" / "estimate-fee" =>
       implicit val decoder: Decoder[DataUpdate] = dataApplication.dataDecoder
       req
         .asR[DataUpdate] { update =>
-          dataApplication.validateUpdate(update).flatMap {
-            case Invalid(e) => BadRequest(InvalidDataUpdate(e.toString).toApplicationError)
-            case _ =>
-              lastGlobalSnapshotStorage.getOrdinal.flatMap {
-                case None => InternalServerError(GL0SnapshotOrdinalUnavailable.toApplicationError)
-                case Some(ord) =>
-                  dataApplication
-                    .estimateFee(ord)(update)
-                    .flatMap { estimatedFee =>
-                      EstimatedFee
-                        .getUpdateHash(update, dataApplication.serializeUpdate)
-                        .map(EstimatedFeeResponse(estimatedFee, _))
-                    }
-                    .flatMap(response => Ok(response.asJson.dropNullValues))
-              }
+          validateDataUpdate(update) {
+            lastGlobalSnapshotStorage.getOrdinal.flatMap {
+              case None => InternalServerError(GL0SnapshotOrdinalUnavailable.toApplicationError)
+              case Some(ordinal) =>
+                dataApplication
+                  .estimateFee(ordinal)(update)
+                  .flatMap { estimatedFee =>
+                    EstimatedFee
+                      .getUpdateHash(update, dataApplication.serializeUpdate)
+                      .map(EstimatedFeeResponse(estimatedFee, _))
+                  }
+                  .map(_.asJson.dropNullValues)
+                  .flatMap(Ok(_))
+            }
           }
         }
         .handleUnknownError
 
     case GET -> Root / "data" =>
-      implicit val signedEncoder: Encoder[Signed[DataUpdate]] = Signed.encoder(dataApplication.dataEncoder)
-      dataUpdatesQueue.view.flatMap(Ok(_))
+      implicit val dataUpdateEncoder: Encoder[DataUpdate] = dataApplication.dataEncoder
+      dataTransactionsQueue.view.flatMap(Ok(_))
 
     case req @ POST -> Root / "data" =>
-      implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
-      req
-        .asR[Signed[DataUpdate]] {
-          _.toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _).flatMap {
-            case Left(_) => BadRequest(InvalidSignature.toApplicationError)
-            case Right(hashed) =>
-              validate(hashed.signed) {
-                dataUpdatesQueue.offer(hashed.signed) >> Ok(("hash" ->> hashed.hash.value) :: HNil)
+      def handleDataUpdates(dataResponse: DataResponse): F[Response[F]] = {
```

Comment:
`handleDataUpdates` receives the data decoded by request decoder, so input is not a `DataResponse`.

Reply:
Currently the requestDecoder returns:
` def requestDecoder(req: Request[F])(implicit f: Async[F]): F[DataResponse]`

Reply:
yeah, so it's wrong. it's not a response yet, so why `DataResponse`? it's a decoded input